src/bun.js/bindings/ZigGlobalObject.cpp

```diff
@@ -1338,6 +1338,14 @@ JSC_DEFINE_HOST_FUNCTION(functionSetTimeout,
         return JSC::JSValue::encode(JSC::JSValue {});
     }
 
+    if (UNLIKELY(!num.isNumber())) {
```

Comment:
you can call setTimeout without a number as the param.

`setTimeout(() => console.log('called'), '1000')`

i forget what `undefined` as the timer value actually ends up doing.

the `JSC__JSValue__coerceToInt32` function is what was used previously by .coerce(i32), so to be safe i think we can just use that.

Reply:
Thanks! I think we should coerce to double, apply ceil(), then cast to int32_t?

Reply:
Instead of getting too funky with ceiling and casting. I think we should narrow down more closely to  what WebKit/Browsers do with setTimeout and simply do that. Via https://developer.mozilla.org/en-US/docs/Web/API/setTimeout#maximum_delay_value values are coerced to Int32 and even roll-over when too high. I assume using the `JSC__JSValue__coerceToInt32` will do that since it uses [`JSValue::toInt32`](https://github.com/WebKit/WebKit/blob/b9f025ad07776f25423540680b468156460a3039/Source/JavaScriptCore/runtime/JSCJSValueInlines.h#L45-L52) under the hood.

Reply:
I think beyond conforming to the expected known behavior of `setTimeout` any other ceiling or rounding should be skipped unless we change the expectation of `bun.sleep`.

Reply:
I'm not sure I fully understand. Do you want to behave in the manner that browsers do, or do you want to skip rounding because everything is done with high precision?

Also, do you want to remove Bun.sleep(Date) or fix it?

Here's a branch that does the minimum fix to get the tests to pass, adds one, and leaves it at that:

https://github.com/oven-sh/bun/compare/main...argosphil:bun:issue-8834-003?expand=1

Reply:
@argosphil Since `Bun.sleep()` is `setTimeout` wrapped in a promise we are dealing with the several factors traditionally. I know that things are different in Bun than a browser because we tie-into event loops and not necessariy the event loop of a browser but just to get the point across traditionally we have to deal with delays being less precise because

* [sometimes delays are throttled](https://developer.mozilla.org/en-US/docs/Web/API/setTimeout#reasons_for_delays_longer_than_specified) to 4ms or so
* inactive tabs (min timeout is 1 second or 15minutes on mobile)
* timeouts are placed in a [task queue](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide#tasks) different than the promise which is in a [microtask queue](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide#tasks_vs._microtasks)

This is to say delays can and do happen (esp relying on both the task and microtask queue) so delays are not precise. See details on [browsers event loop](https://developer.mozilla.org/en-US/docs/Web/API/HTML_DOM_API/Microtask_guide/In_depth) and [Node's event loop](https://nodejs.org/en/guides/event-loop-timers-and-nexttick/) (also on [Promises in Node's event loop](https://nodejs.org/en/learn/asynchronous-work/understanding-setimmediate)) which Bun simulates

> Do you want to behave in the manner that browsers do, or do you want to skip rounding because everything is done with high precision?

Do nothing if we behave as passing off to `setTimeout` wrapped in a promise. If we differ from that and our `setTimeout` implementation needs tweaking then that is where we can tweak.

Reply:
@jdalton Thanks for your response!

> @argosphil Since `Bun.sleep()` is `setTimeout` wrapped in a promise

I agree. I proposed to fix the one exception (setTimeout doesn't take a Date object, Bun.sleep does) first, then look into switching all timeouts to high precision.

There is one point where we do and should continue to deviate from the HTML Standard, which is that it mandates a 4 ms delay after five nested calls. Do you think we should implement that delay?

Note that high precision does not violate the HTML Standard. I do not know to what extent it behaves differently from what node does, though.

> Do nothing if we behave as passing off to `setTimeout` wrapped in a promise. If we differ from that and our `setTimeout` implementation needs tweaking then that is where we can tweak.

Yes, you've convinced me of that point, which is why I think I should open a PR for the issue-8834-003 branch, which doesn't touch setTimeout or Bun.sleep(number) in any way.

Reply:
@argosphil 
> There is one point where we do and should continue to deviate from the HTML Standard, which is that it mandates a 4 ms delay after five nested calls. Do you think we should implement that delay?

I'd look to if Node does anything there. Our compat burden is for the Node behavior.

> Yes, you've convinced me of that point, which is why I think I should open a PR for the issue-8834-003 branch, which doesn't touch setTimeout or Bun.sleep(number) in any way.

Awesome! Thank you @argosphil  for digging into this!